design/baremetal-operator/hostdata_crd.md

```diff
@@ -0,0 +1,232 @@
+the same name & namespace will be used for HostData as its owning object BareMetalHost.
+The spec of HostData will contain the same fields & hardware information as currently
+under `status.hardware` of BareMetalHost.
+
+OwnerReferences refers to the linked BareMetalHost and is added by the operator.
+To see all the available fields of ObjectReference type, check [core/v1](https://pkg.go.dev/k8s.io/api/core/v1#ObjectReference) package.
+
+There will be duplication of inspection data, i.e.,in the HostData spec and BareMetalHost
+status, for the purpose of avoiding breaking API changes. However, we plan to drop
+[hardware](https://github.com/metal3-io/baremetal-operator/blob/05d12b6768a9989a9a4e61dad6cd1f9e84a6e078/apis/metal3.io/v1alpha1/baremetalhost_types.go#L721)
+field from the status of BareMetalHost in the first version bump of the API.
+
+### Pivoting
+
+To pivot objects excluded from Cluster API chain to a target cluster, one can set
+`clusterctl.cluster.x-k8s.io=""` label on HostData and BareMetalHost CRDs.
+Current status annotation based pivoting workflow will stay as it is. However,
+CAPM3 machine controller will no longer copy `status.hardware` of BareMetalHost
+to the annotation. Because that part will be avialable in the form of HostData CR
+in the target cluster.
+
+### Implementation Details/Notes/Constraints
+
+The life of HostData is dependent on BareMetalHost. When BareMetalHost is entering
+inspection state, the HostData will be created. Failure in reaching inspection state
+or having `inspect.metal3.io: disabled` (without `inspect.metal3.io/hardwaredetails`
+annotation) annotation on BareMetalHost will force the operator to not create the
+HostData. Deleting a BareMetalHost will result in deletion of a HostData respectively.
+
+HostData should be immutable and this will be controlled via [Validating Webhook](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#validatingadmissionwebhook) and with RBAC permissions.
+In case, when [re-inspection](https://github.com/metal3-io/baremetal-operator/blob/master/docs/inspectAnnotation.md) of a `available` BareMetalHost is requested,
+new HostData will be re-created, because updating the old HostData will be
+blocked by the Validating Webhook.
+
+Deprovisioning of the BareMetalHost must not delete the HostData. It can
+be deleted either when BareMetalHost is being deleted or when re-inspection
+is requested. To ensure HostData availability while BareMetalHost is provisioned,
+a finalizer should be set on the HostData to prevent its accidental deletion.
+
+Only operator will be granted with create permissions for HostData and everyone
+else will be given no write permissions. If the BareMetalHost was inspected but
+HostData is missing, operator acts differently depending on the current
+state of the BareMetalHost.
+
+- host is provisoned: enter error state
+- host is available:
+    1. copy inspection data from the status of BareMetalHost if exists;
```

Comment:
I'm still not convinced about point 1), and think we should keep just 2). It could have a couple of benefits:
a) HD cr will be effectively the only trusted source for haedware details
b) When in future the `status.hardware` field will be removed, the user will not observe a different behavior for that scenario.
Anyhow, I'd like to hear the other opinions on that.

Reply:
@zaneb I would like to hear your opinion too on this since you also suggested earlier about bi-directional copy in https://github.com/metal3-io/metal3-docs/pull/208#discussion_r742196355.

Reply:
I'd have thought we'd need this for backward compatibility, at least initially. CAPM3 currently stores the hardwaredetails in the status annotation during a pivot. OpenShift installer does something similar (for control plane nodes). I don't know if there are any other third parties relying on the status annotation, but if we don't use it when it is available and the resource is not, we're requiring all of them to implement support for the new CRD before we have even landed the code to use it.

Reply:
Ok for backward compatibility. @fmuyassarov maybe in the implementation PR we could add a specific log trace when performing such back copy (bmh.status -> hd) in case of missing resource for warning the user, for clarity (even though here in the design doc it's mentioned that in future only re-inspection will be the viable option).

Reply:
@andfasano +1 for that. And when we are going remove the status completely, we can clean up the tracking also accordingly.